### PR TITLE
New Resource - `azurerm_hpc_cache_blob_nfs_target`

### PR DESCRIPTION
**Pull request by magodo**
This PR implements a new HPC resource: `azurerm_hpc_cache_blob_nfs_target`. Note that due to the level of NFS support of storage account, the test can be only passed in limited regions (incl. west europe, eastus, etc.).

## Test Result

```bash
terraform-provider-azurerm on  hpc_blobnfs_target via 🐹 v1.16.2
💢 TF_ACC=1 go test -parallel 2 -v -timeout=8h ./azurerm/internal/services/hpccache -run="TestAccHPCCacheBlobNFSTarget_update"
2021/05/12 11:13:55 [DEBUG] not using binary driver name, it's no longer needed
2021/05/12 11:13:55 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccHPCCacheBlobNFSTarget_update
=== PAUSE TestAccHPCCacheBlobNFSTarget_update
=== CONT  TestAccHPCCacheBlobNFSTarget_update
--- PASS: TestAccHPCCacheBlobNFSTarget_update (1912.01s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hpccache    1912.138s

terraform-provider-azurerm on  hpc_blobnfs_target via 🐹 v1.16.2 took 31m58s
💤 TF_ACC=1 go test -parallel 2 -v -timeout=8h ./azurerm/internal/services/hpccache -run="TestAccHPCCacheBlobNFSTarget_accessPolicy"
2021/05/12 12:52:52 [DEBUG] not using binary driver name, it's no longer needed
2021/05/12 12:52:52 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccHPCCacheBlobNFSTarget_accessPolicy
=== PAUSE TestAccHPCCacheBlobNFSTarget_accessPolicy
=== CONT  TestAccHPCCacheBlobNFSTarget_accessPolicy
--- PASS: TestAccHPCCacheBlobNFSTarget_accessPolicy (2852.94s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hpccache    2853.118s

terraform-provider-azurerm on  hpc_blobnfs_target via 🐹 v1.16.2 took 47m40s
💤 TF_ACC=1 go test -parallel 2 -v -timeout=8h ./azurerm/internal/services/hpccache -run="TestAccHPCCacheBlobNFSTarget_basic"
2021/05/12 14:00:15 [DEBUG] not using binary driver name, it's no longer needed
2021/05/12 14:00:15 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccHPCCacheBlobNFSTarget_basic
=== PAUSE TestAccHPCCacheBlobNFSTarget_basic
=== CONT  TestAccHPCCacheBlobNFSTarget_basic
--- PASS: TestAccHPCCacheBlobNFSTarget_basic (2184.23s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hpccache    2184.385s

terraform-provider-azurerm on  hpc_blobnfs_target via 🐹 v1.16.2 took 36m31s
💤 TF_ACC=1 go test -parallel 2 -v -timeout=8h ./azurerm/internal/services/hpccache -run="TestAccHPCCacheBlobNFSTarget_requiresImport"
2021/05/12 14:36:53 [DEBUG] not using binary driver name, it's no longer needed
2021/05/12 14:36:53 [DEBUG] not using binary driver name, it's no longer needed
=== RUN   TestAccHPCCacheBlobNFSTarget_requiresImport
=== PAUSE TestAccHPCCacheBlobNFSTarget_requiresImport
=== CONT  TestAccHPCCacheBlobNFSTarget_requiresImport
--- PASS: TestAccHPCCacheBlobNFSTarget_requiresImport (1968.57s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hpccache    1968.726s
```

### CODE REVIEW

File: azurerm/internal/services/hpccache/hpc_cache_blob_nfs_target_resource.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+		{
+			NamespacePath:   &namespacePath,
+			TargetPath:      utils.String("/"),
+			NfsExport:       utils.String("/"),
+			NfsAccessPolicy: utils.String(d.Get("access_policy_name").(string)),
+		},
+	}
+	param := &storagecache.StorageTarget{
+		StorageTargetProperties: &storagecache.StorageTargetProperties{
+			Junctions:  &namespaceJunction,
```
Comment (katbyte) [inline azurerm/internal/services/hpccache/hpc_cache_blob_nfs_target_resource.go:138]: should we just inline this instead of creating a var for it?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @magodo - overall this looks good aside from a couple comments. running the tests i'm seeing failures thou:

```
------- Stdout: -------
=== RUN   TestAccHPCCacheBlobNFSTarget_accessPolicy
=== PAUSE TestAccHPCCacheBlobNFSTarget_accessPolicy
=== CONT  TestAccHPCCacheBlobNFSTarget_accessPolicy
    testing.go:620: Step 1/8 error: Error running apply: exit status 1
        
        Error: Error waiting for HPC Cache "acctest-HPCC-210519233234713607" (Resource Group "acctestRG-storage-210519233234713607") to finish provisioning: Code="CreateResourceTimeout" Message="Timed out creating Azure resources.   Please delete the cache and try again."
        
          with azurerm_hpc_cache.test,
          on terraform_plugin_test.tf line 134, in resource "azurerm_hpc_cache" "test":
         134: resource "azurerm_hpc_cache" "test" {
        
    testing_new.go:21: WARNING: destroy failed, so remote objects may still exist and be subject to billing
    testing_new.go:21: failed to destroy: exit status 1
        
        Error: deleting Subnet: (Name "acctestsub-210519233234713607" / Virtual Network Name "acctest-VN-210519233234713607" / Resource Group "acctestRG-storage-210519233234713607"): network.SubnetsClient#Delete: Failure sending request: StatusCode=0 -- Original Error: Code="InUseSubnetCannotBeDeleted" Message="Subnet acctestsub-210519233234713607 is in use by /subscriptions/*******/resourceGroups/ACCTESTRG-STORAGE-210519233234713607/providers/Microsoft.Network/networkInterfaces/prod-8ab244be-172b-4ee1-84ad-ca72a7c1562e-cluster-nic-0/ipConfigurations/primary and cannot be deleted. In order to delete the subnet, delete all the resources within the subnet. See aka.ms/deletesubnet." Details=[]
        
   ------- Stdout: -------
=== RUN   TestAccHPCCacheBlobNFSTarget_basic
=== PAUSE TestAccHPCCacheBlobNFSTarget_basic
=== CONT  TestAccHPCCacheBlobNFSTarget_basic
    testing.go:620: Step 1/2 error: Error running apply: exit status 1
        
        Error: Error waiting for Storage Target: (Name "acctest-HPCCTGT-puzqr" / Cache Name "acctest-HPCC-210519233234668059" / Resource Group "acctestRG-storage-210519233234668059"): Code="AscInternalError" Message="Error encountered deploying the storage target: 'acctest-HPCCTGT-puzqr'."
        
          with azurerm_hpc_cache_blob_nfs_target.test,
          on terraform_plugin_test.tf line 144, in resource "azurerm_hpc_cache_blob_nfs_target" "test":
         144: resource "azurerm_hpc_cache_blob_nfs_target" "test" {
        
--- FAIL: TestAccHPCCacheBlobNFSTarget_basic (1360.13s)
FAIL

------- Stderr: -------
2021/05/19 23:32:34 [DEBUG] not using binary driver name, it's no longer needed
2021/05/19 23:32:34 [DEBUG] not using binary driver name, it's no longer needed

------- Stdout: -------
=== RUN   TestAccHPCCacheBlobNFSTarget_requiresImport
=== PAUSE TestAccHPCCacheBlobNFSTarget_requiresImport
=== CONT  TestAccHPCCacheBlobNFSTarget_requiresImport
    testing.go:620: Step 1/2 error: Error running apply: exit status 1
        
        Error: Error creating Storage Target: (Name "acctest-HPCCTGT-rhfbg" / Cache Name "acctest-HPCC-210519235513184265" / Resource Group "acctestRG-storage-210519235513184265"): storagecache.StorageTargetsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidParameter" Message="Target storage account or container does not exist or hasn't sufficient permissions." Target="storageTarget.blobNfs.target"
        
          with azurerm_hpc_cache_blob_nfs_target.test,
          on terraform_plugin_test.tf line 144, in resource "azurerm_hpc_cache_blob_nfs_target" "test":
         144: resource "azurerm_hpc_cache_blob_nfs_target" "test" {
        
--- FAIL: TestAccHPCCacheBlobNFSTarget_requiresImport (1353.34s)
FAIL

------- Stderr: -------
2021/05/19 23:55:12 [DEBUG] not using binary driver name, it's no longer needed
2021/05/19 23:55:13 [DEBUG] not using binary driver name, it's no longer needed
```

File: azurerm/internal/services/hpccache/hpc_cache_blob_nfs_target_resource.go (hunk 1 above)
Comment (magodo) [reply azurerm/internal/services/hpccache/hpc_cache_blob_nfs_target_resource.go:138]: The `namespaceJunction` is a slice, for which we do not have a utility function to get its pointer. That's the reason to declare a local var here.

**Review by katbyte — CHANGES_REQUESTED** [review]
still have a test failure:
```
------- Stdout: -------
=== RUN   TestAccHPCCacheBlobNFSTarget_requiresImport
=== PAUSE TestAccHPCCacheBlobNFSTarget_requiresImport
=== CONT  TestAccHPCCacheBlobNFSTarget_requiresImport
    testing.go:620: Step 1/2 error: Error running apply: exit status 1
        
        Error: Error waiting for Storage Target: (Name "acctest-HPCCTGT-7gtxo" / Cache Name "acctest-HPCC-210601181157052404" / Resource Group "acctestRG-storage-210601181157052404"): Code="InternalExecutionError" Message="An internal execution error occurred."
        
          with azurerm_hpc_cache_blob_nfs_target.test,
          on terraform_plugin_test.tf line 115, in resource "azurerm_hpc_cache_blob_nfs_target" "test":
         115: resource "azurerm_hpc_cache_blob_nfs_target" "test" {
        
--- FAIL: TestAccHPCCacheBlobNFSTarget_requiresImport (1740.02s)
FAIL

------- Stderr: -------
2021/06/01 18:11:56 [DEBUG] not using binary driver name, it's no longer needed
2021/06/01 18:11:56 [DEBUG] not using binary driver name, it's no longer needed
```

**Comment by magodo** [comment]
@katbyte I'm still discussing with the service team for this kind of random failures. I'm wondering whether we can merge this PR with these intermittent issue existing, since these are issues on service side.

**Comment by katbyte** [comment]
@magodo i've not been able to get a clean run in TC and it seems this is a constant failure now:
```
------- Stdout: -------
=== RUN   TestAccHPCCacheBlobNFSTarget_requiresImport
=== PAUSE TestAccHPCCacheBlobNFSTarget_requiresImport
=== CONT  TestAccHPCCacheBlobNFSTarget_requiresImport
    testing.go:620: Step 1/2 error: Error running apply: exit status 1
        
        Error: Error creating Storage Target: (Name "acctest-HPCCTGT-w0gd3" / Cache Name "acctest-HPCC-210603165047909801" / Resource Group "acctestRG-storage-210603165047909801"): storagecache.StorageTargetsClient#CreateOrUpdate: Failure sending request: StatusCode=0 -- Original Error: Code="InvalidParameter" Message="Target storage account or container does not exist or hasn't sufficient permissions." Target="storageTarget.blobNfs.target"
        
          with azurerm_hpc_cache_blob_nfs_target.test,
          on terraform_plugin_test.tf line 115, in resource "azurerm_hpc_cache_blob_nfs_target" "test":
         115: resource "azurerm_hpc_cache_blob_nfs_target" "test" {
        
--- FAIL: TestAccHPCCacheBlobNFSTarget_requiresImport (1347.75s)
FAIL

------- Stderr: -------
2021/06/03 16:50:47 [DEBUG] not using binary driver name, it's no longer needed
2021/06/03 16:50:47 [DEBUG] not using binary driver name, it's no longer needed

```

**Comment by magodo** [comment]
@katbyte The service team is still looking into these errors.

**Review by katbyte — APPROVED** [review]
LGTM 👍

**Pull request merged**